--- ../src-base/minecraft/net/minecraft/pathfinding/Path.java
+++ ../src-work/minecraft/net/minecraft/pathfinding/Path.java
@@ -46,7 +46,9 @@
 
     public void setPoint(int index, PathPoint point)
     {
-        this.points[index] = point;
+        synchronized (points) {
+            this.points[index] = point;
+        }
     }
 
     public int getCurrentPathLength()
@@ -69,11 +71,13 @@
         this.currentPathIndex = currentPathIndexIn;
     }
 
-    public Vec3d getVectorFromIndex(Entity entityIn, int index)
-    {
-        double d0 = (double)this.points[index].x + (double)((int)(entityIn.width + 1.0F)) * 0.5D;
-        double d1 = (double)this.points[index].y;
-        double d2 = (double)this.points[index].z + (double)((int)(entityIn.width + 1.0F)) * 0.5D;
+    public Vec3d getVectorFromIndex(Entity entityIn, int index) {
+        if (index > points.length) {
+            return new Vec3d(entityIn.posX, entityIn.posY, entityIn.posZ);
+        }
+        double d0 = (double) this.points[index].x + (double) ((int) (entityIn.width + 1.0F)) * 0.5D;
+        double d1 = (double) this.points[index].y;
+        double d2 = (double) this.points[index].z + (double) ((int) (entityIn.width + 1.0F)) * 0.5D;
         return new Vec3d(d0, d1, d2);
     }
 
@@ -82,28 +86,22 @@
         return this.getVectorFromIndex(entityIn, this.currentPathIndex);
     }
 
-    public Vec3d getCurrentPos()
-    {
-        PathPoint pathpoint = this.points[this.currentPathIndex];
-        return new Vec3d((double)pathpoint.x, (double)pathpoint.y, (double)pathpoint.z);
+    public Vec3d getCurrentPos() {
+        PathPoint pathpoint;
+        synchronized (points) {
+            pathpoint = this.points[this.currentPathIndex];
+        }
+        return new Vec3d((double) pathpoint.x, (double) pathpoint.y, (double) pathpoint.z);
     }
 
-    public boolean isSamePath(Path pathentityIn)
-    {
-        if (pathentityIn == null)
-        {
+    public boolean isSamePath(Path pathentityIn) {
+        if (pathentityIn == null) {
             return false;
-        }
-        else if (pathentityIn.points.length != this.points.length)
-        {
+        } else if (pathentityIn.points.length != this.points.length) {
             return false;
-        }
-        else
-        {
-            for (int i = 0; i < this.points.length; ++i)
-            {
-                if (this.points[i].x != pathentityIn.points[i].x || this.points[i].y != pathentityIn.points[i].y || this.points[i].z != pathentityIn.points[i].z)
-                {
+        } else {
+            for (int i = 0; i < this.points.length; ++i) {
+                if (this.points[i].x != pathentityIn.points[i].x || this.points[i].y != pathentityIn.points[i].y || this.points[i].z != pathentityIn.points[i].z) {
                     return false;
                 }
             }
