--- ../src-base/minecraft/net/minecraft/pathfinding/Path.java
+++ ../src-work/minecraft/net/minecraft/pathfinding/Path.java
@@ -7,152 +7,132 @@
 import net.minecraftforge.fml.relauncher.Side;
 import net.minecraftforge.fml.relauncher.SideOnly;
 
-public class Path
-{
-    private final PathPoint[] points;
+import java.util.concurrent.atomic.AtomicInteger;
+import java.util.concurrent.atomic.AtomicReferenceArray;
+
+public class Path {
+    private final AtomicReferenceArray<PathPoint> points;
     private PathPoint[] openSet = new PathPoint[0];
     private PathPoint[] closedSet = new PathPoint[0];
     @SideOnly(Side.CLIENT)
     private PathPoint target;
-    private int currentPathIndex;
-    private int pathLength;
+    private final AtomicInteger currentPathIndex = new AtomicInteger(0);
+    private final AtomicInteger pathLength = new AtomicInteger(0);
 
-    public Path(PathPoint[] pathpoints)
-    {
-        this.points = pathpoints;
-        this.pathLength = pathpoints.length;
+    public Path(PathPoint[] pathpoints) {
+        this.points = new AtomicReferenceArray<>(pathpoints);
+        this.pathLength.set(pathpoints.length);
     }
 
-    public void incrementPathIndex()
-    {
-        ++this.currentPathIndex;
+    public void incrementPathIndex() {
+        this.currentPathIndex.getAndIncrement();
     }
 
-    public boolean isFinished()
-    {
-        return this.currentPathIndex >= this.pathLength;
+    public boolean isFinished() {
+        return this.currentPathIndex.get() >= this.pathLength.get();
     }
 
     @Nullable
-    public PathPoint getFinalPathPoint()
-    {
-        return this.pathLength > 0 ? this.points[this.pathLength - 1] : null;
+    public PathPoint getFinalPathPoint() {
+        return this.pathLength.get() > 0 ? this.points.get(this.pathLength.get() - 1) : null;
     }
 
-    public PathPoint getPathPointFromIndex(int index)
-    {
-        return this.points[index];
+    public PathPoint getPathPointFromIndex(int index) {
+        return this.points.get(index);
     }
 
-    public void setPoint(int index, PathPoint point)
-    {
-        this.points[index] = point;
+    public void setPoint(int index, PathPoint point) {
+        this.points.set(index, point);
     }
 
-    public int getCurrentPathLength()
-    {
-        return this.pathLength;
+    public int getCurrentPathLength() {
+        return this.pathLength.get();
     }
 
-    public void setCurrentPathLength(int length)
-    {
-        this.pathLength = length;
+    public void setCurrentPathLength(int length) {
+        this.pathLength.set(length);
     }
 
-    public int getCurrentPathIndex()
-    {
-        return this.currentPathIndex;
+    public int getCurrentPathIndex() {
+        return this.currentPathIndex.get();
     }
 
-    public void setCurrentPathIndex(int currentPathIndexIn)
-    {
-        this.currentPathIndex = currentPathIndexIn;
+    public void setCurrentPathIndex(int currentPathIndexIn) {
+        this.currentPathIndex.set(currentPathIndexIn);
     }
 
-    public Vec3d getVectorFromIndex(Entity entityIn, int index)
-    {
-        double d0 = (double)this.points[index].x + (double)((int)(entityIn.width + 1.0F)) * 0.5D;
-        double d1 = (double)this.points[index].y;
-        double d2 = (double)this.points[index].z + (double)((int)(entityIn.width + 1.0F)) * 0.5D;
+    public Vec3d getVectorFromIndex(Entity entityIn, int index) {
+        if (index > points.length()) {
+            return new Vec3d(entityIn.posX, entityIn.posY, entityIn.posZ);
+        }
+        double d0 = (double) this.points.get(index).x + (double) ((int) (entityIn.width + 1.0F)) * 0.5D;
+        double d1 = this.points.get(index).y;
+        double d2 = (double) this.points.get(index).z + (double) ((int) (entityIn.width + 1.0F)) * 0.5D;
         return new Vec3d(d0, d1, d2);
+
     }
 
-    public Vec3d getPosition(Entity entityIn)
-    {
-        return this.getVectorFromIndex(entityIn, this.currentPathIndex);
+    public Vec3d getPosition(Entity entityIn) {
+        return this.getVectorFromIndex(entityIn, this.currentPathIndex.get());
     }
 
-    public Vec3d getCurrentPos()
-    {
-        PathPoint pathpoint = this.points[this.currentPathIndex];
-        return new Vec3d((double)pathpoint.x, (double)pathpoint.y, (double)pathpoint.z);
+    public Vec3d getCurrentPos() {
+        PathPoint pathpoint;
+        pathpoint = this.points.get(this.currentPathIndex.get());
+        return new Vec3d(pathpoint.x, pathpoint.y, pathpoint.z);
     }
 
-    public boolean isSamePath(Path pathentityIn)
-    {
-        if (pathentityIn == null)
-        {
+    public boolean isSamePath(Path pathentityIn) {
+        if (pathentityIn == null) {
             return false;
-        }
-        else if (pathentityIn.points.length != this.points.length)
-        {
+        } else if (pathentityIn.points.length() != this.points.length()) {
             return false;
-        }
-        else
-        {
-            for (int i = 0; i < this.points.length; ++i)
-            {
-                if (this.points[i].x != pathentityIn.points[i].x || this.points[i].y != pathentityIn.points[i].y || this.points[i].z != pathentityIn.points[i].z)
-                {
+        } else {
+            for (int i = 0; i < this.points.length(); ++i) {
+                if (this.points.get(i).x != pathentityIn.points.get(i).x || this.points.get(i).y != pathentityIn.points.get(i).y || this.points.get(i).z != pathentityIn.points.get(i).z) {
                     return false;
                 }
             }
 
             return true;
         }
+
     }
 
     @SideOnly(Side.CLIENT)
-    public PathPoint[] getOpenSet()
-    {
+    public PathPoint[] getOpenSet() {
         return this.openSet;
     }
 
     @SideOnly(Side.CLIENT)
-    public PathPoint[] getClosedSet()
-    {
+    public PathPoint[] getClosedSet() {
         return this.closedSet;
     }
 
     @SideOnly(Side.CLIENT)
-    public PathPoint getTarget()
-    {
+    public PathPoint getTarget() {
         return this.target;
     }
 
     @SideOnly(Side.CLIENT)
-    public static Path read(PacketBuffer buf)
-    {
+    public static Path read(PacketBuffer buf) {
         int i = buf.readInt();
         PathPoint pathpoint = PathPoint.createFromBuffer(buf);
         PathPoint[] apathpoint = new PathPoint[buf.readInt()];
 
-        for (int j = 0; j < apathpoint.length; ++j)
-        {
+        for (int j = 0; j < apathpoint.length; ++j) {
             apathpoint[j] = PathPoint.createFromBuffer(buf);
         }
 
         PathPoint[] apathpoint1 = new PathPoint[buf.readInt()];
 
-        for (int k = 0; k < apathpoint1.length; ++k)
-        {
+        for (int k = 0; k < apathpoint1.length; ++k) {
             apathpoint1[k] = PathPoint.createFromBuffer(buf);
         }
 
         PathPoint[] apathpoint2 = new PathPoint[buf.readInt()];
 
-        for (int l = 0; l < apathpoint2.length; ++l)
-        {
+        for (int l = 0; l < apathpoint2.length; ++l) {
             apathpoint2[l] = PathPoint.createFromBuffer(buf);
         }
 
@@ -160,7 +140,7 @@
         path.openSet = apathpoint1;
         path.closedSet = apathpoint2;
         path.target = pathpoint;
-        path.currentPathIndex = i;
+        path.currentPathIndex.set(i);
         return path;
     }
 }
