--- ../src-base/minecraft/net/minecraft/pathfinding/PathNavigate.java
+++ ../src-work/minecraft/net/minecraft/pathfinding/PathNavigate.java
@@ -22,11 +22,11 @@
     @Nullable
     protected Path currentPath;
     protected double speed;
-    private final IAttributeInstance pathSearchRange;
+    public final IAttributeInstance pathSearchRange;
     protected int totalTicks;
     private int ticksAtLastPos;
-    private Vec3d lastPosCheck = Vec3d.ZERO;
-    private Vec3d timeoutCachedNode = Vec3d.ZERO;
+    public Vec3d lastPosCheck = Vec3d.ZERO;
+    public Vec3d timeoutCachedNode = Vec3d.ZERO;
     private long timeoutTimer;
     private long lastTimeoutCheck;
     private double timeoutLimit;
@@ -257,16 +257,18 @@
         int k = MathHelper.ceil(this.entity.width);
         int l = MathHelper.ceil(this.entity.height);
         int i1 = k;
-
-        for (int j1 = i - 1; j1 >= this.currentPath.getCurrentPathIndex(); --j1)
-        {
-            if (this.isDirectPathBetweenPoints(vec3d, this.currentPath.getVectorFromIndex(this.entity, j1), k, l, i1))
-            {
-                this.currentPath.setCurrentPathIndex(j1);
-                break;
+        if (this.currentPath!=null) {
+            for (int j1 = i - 1; j1 >= this.currentPath.getCurrentPathIndex(); --j1) {
+                Vec3d vec = this.currentPath.getVectorFromIndex(this.entity, j1);
+                if (vec == null) {
+                    continue;
+                }
+                if (this.isDirectPathBetweenPoints(vec3d, vec, k, l, i1)) {
+                    this.currentPath.setCurrentPathIndex(j1);
+                    break;
+                }
             }
         }
-
         this.checkForStuck(vec3d);
     }
 
