--- ../src-base/minecraft/net/minecraft/command/ServerCommandManager.java
+++ ../src-work/minecraft/net/minecraft/command/ServerCommandManager.java
@@ -1,31 +1,8 @@
 package net.minecraft.command;
 
-import net.minecraft.command.server.CommandBanIp;
-import net.minecraft.command.server.CommandBanPlayer;
-import net.minecraft.command.server.CommandBroadcast;
-import net.minecraft.command.server.CommandDeOp;
-import net.minecraft.command.server.CommandEmote;
-import net.minecraft.command.server.CommandListBans;
-import net.minecraft.command.server.CommandListPlayers;
-import net.minecraft.command.server.CommandMessage;
-import net.minecraft.command.server.CommandMessageRaw;
-import net.minecraft.command.server.CommandOp;
-import net.minecraft.command.server.CommandPardonIp;
-import net.minecraft.command.server.CommandPardonPlayer;
-import net.minecraft.command.server.CommandPublishLocalServer;
-import net.minecraft.command.server.CommandSaveAll;
-import net.minecraft.command.server.CommandSaveOff;
-import net.minecraft.command.server.CommandSaveOn;
-import net.minecraft.command.server.CommandScoreboard;
-import net.minecraft.command.server.CommandSetBlock;
-import net.minecraft.command.server.CommandSetDefaultSpawnpoint;
-import net.minecraft.command.server.CommandStop;
-import net.minecraft.command.server.CommandSummon;
-import net.minecraft.command.server.CommandTeleport;
-import net.minecraft.command.server.CommandTestFor;
-import net.minecraft.command.server.CommandTestForBlock;
-import net.minecraft.command.server.CommandWhitelist;
+import net.minecraft.command.server.*;
 import net.minecraft.entity.player.EntityPlayer;
+import net.minecraft.entity.player.EntityPlayerMP;
 import net.minecraft.network.rcon.RConConsoleSource;
 import net.minecraft.server.MinecraftServer;
 import net.minecraft.tileentity.CommandBlockBaseLogic;
@@ -40,6 +17,36 @@
     public ServerCommandManager(MinecraftServer serverIn)
     {
         this.server = serverIn;
+        CommandBase.setCommandListener(this);
+    }
+    public static class CommandForceKick extends CommandBase {
+        @Override
+        public String getName() {
+            return "forcekick";
+        }
+
+        @Override
+        public String getUsage(ICommandSender sender) {
+            return "/forcekick <player>";
+        }
+
+        @Override
+        public void execute(MinecraftServer server, ICommandSender sender, String[] args) throws CommandException {
+            if (args.length > 0 && args[0].length() > 1) {
+                EntityPlayerMP entityplayermp = server.getPlayerList().getPlayerByUsername(args[0]);
+                if(entityplayermp!=null) {
+                    entityplayermp.connection.getNetworkManager().forceDisconnect();
+                }
+            }
+        }
+
+        @Override
+        public int getRequiredPermissionLevel()
+        {
+            return 3;
+        }
+    }
+    public void registerVanillaCommands() {
         this.registerCommand(new CommandTime());
         this.registerCommand(new CommandGameMode());
         this.registerCommand(new CommandDifficulty());
@@ -57,6 +64,7 @@
         this.registerCommand(new CommandEnchant());
         this.registerCommand(new CommandParticle());
         this.registerCommand(new CommandEmote());
+        this.registerCommand(new CommandForceKick());
         this.registerCommand(new CommandShowSeed());
         this.registerCommand(new CommandHelp());
         this.registerCommand(new CommandDebug());
@@ -89,8 +97,7 @@
         this.registerCommand(new CommandLocate());
         this.registerCommand(new CommandReload());
         this.registerCommand(new CommandFunction());
-
-        if (serverIn.isDedicatedServer())
+        if (server.isDedicatedServer())
         {
             this.registerCommand(new CommandOp());
             this.registerCommand(new CommandDeOp());
@@ -147,7 +154,7 @@
             }
         }
 
-        if (sender != minecraftserver && minecraftserver.worlds[0].getGameRules().getBoolean("logAdminCommands"))
+        if (sender != minecraftserver && minecraftserver.worlds[0].getGameRules().getBoolean("logAdminCommands") && !org.spigotmc.SpigotConfig.silentCommandBlocks) // Spigot
         {
             minecraftserver.sendMessage(itextcomponent);
         }
